pandas/util/_test_decorators.py

```diff
@@ -232,38 +232,67 @@ def documented_fixture(fixture):
     return documented_fixture
 
 
-def check_file_leaks(func) -> Callable:
+class check_file_leaks(ContextDecorator):
     """
-    Decorate a test function to check that we are not leaking file descriptors.
-    """
-    with file_leak_context():
-        return func
-
+    Use psutil and ResourceWarning to identify forgotten resources.
 
-@contextmanager
-def file_leak_context():
-    """
-    ContextManager analogue to check_file_leaks.
+    ResourceWarnings that contain the string 'ssl' are ignored as they are very likely
+    caused by boto3 (GH#17058).
     """
-    psutil = safe_import("psutil")
-    if not psutil:
-        yield
-    else:
-        proc = psutil.Process()
-        flist = proc.open_files()
-        conns = proc.connections()
-
-        yield
-
-        flist2 = proc.open_files()
-        # on some builds open_files includes file position, which we _dont_
-        #  expect to remain unchanged, so we need to compare excluding that
-        flist_ex = [(x.path, x.fd) for x in flist]
-        flist2_ex = [(x.path, x.fd) for x in flist2]
-        assert flist2_ex == flist_ex, (flist2, flist)
 
-        conns2 = proc.connections()
-        assert conns2 == conns, (conns2, conns)
+    def __init__(self, ignore_connections: bool = False):
```

Comment:
why would we ignore connections?

Reply:
I disabled them when extending the test to pandas/tests/io since we probably catch some ssl connections from boto3. I will double check whether that is needed. Tests that were previously covered by check_file_leaks have their connections checked.

Do you have an idea what might be causing https://github.com/pandas-dev/pandas/pull/39047#issuecomment-767301178 ?

Reply:
> Do you have an idea what might be causing #39047 (comment) ?

I've never had any luck in tracking down the unclosed ssl socket warnings.  the connections part of check_file_leaks was supposed to identify them, but apparently doesnt

Reply:
I have a feeling that associating ResoruceWarnings with different tests might be an issue with pytest-xdist. I found a few issues about warnings on their github but nothing that would directly explain what I see in this PR.

Reply:
could try some CI runs with xdist disabled?

Reply:
what is the bes way to disable it? I tried to add `-n 0` to `addopts` in setup.cfg but that has no effect as pytest seems to be called with `-n auto`